lib/spack/spack/environment.py

```diff
@@ -1509,6 +1509,66 @@ def concretized_specs(self):
         for s, h in zip(self.concretized_user_specs, self.concretized_order):
             yield (s, self.specs_by_hash[h])
 
+    def matching_spec(self, spec):
+        """
+        Given a spec (likely not concretized), find a matching concretized
+        spec in the environment.
+
+        The matching spec does not have to be installed in the environment,
+        but must be concrete (specs added with `spack add` without an
+        intervening `spack concretize` will not be matched).
+
+        If there is a single root spec that matches the provided spec or a
+        single dependency spec that matches the provided spec, then the
+        concretized instance of that spec will be returned.
+
+        If multiple root specs match the provided spec, or no root specs match
+        and multiple dependency specs match, then this raises an error
+        and reports all matching specs.
+        """
+        # Root specs will be keyed by concrete spec, value abstract
+        # Dependency-only specs will have value None
+        matches = {}
+
+        for user_spec, concretized_user_spec in self.concretized_specs():
+            if concretized_user_spec.satisfies(spec):
+                matches[concretized_user_spec] = user_spec
+            for dep_spec in concretized_user_spec.traverse(root=False):
+                if dep_spec.satisfies(spec):
+                    # Don't overwrite the abstract spec if present
+                    # If not present already, set to None
+                    matches[dep_spec] = matches.get(dep_spec)
```

Comment:
I think explicitly setting this to `None` would be more clear: I don't think anything depends on the behavior that this starts with `{dep_spec: None}` and then turns into `{dep_spec: dep_spec}` on subsequent iterations.

Reply:
This doesn't turn from `{dep_spec: None}` to `{dep_spec: dep_spec}`.

This line is to handle the case where `dep_spec` is also a root that was already handled. So we're taking one of the following:

`{dep_spec: abstract_spec} -> {dep_spec: abstract_spec}`
`{dep_spec: None} -> {dep_spec: None}`
`{} -> {dep_spec: None}`

So we take the existing value if there is one, otherwise `None`, and that's what we add.

Reply:
Whoops yes that makes sense. I think it would be more straightforward though if you did:

```
if not matches[dep_spec]:
  matches[dep_spec] = None
```

Reply:
That won't work, because that will throw an exception if dep_spec is not already in matches.